rfcs/require-by-string.md

```diff
@@ -0,0 +1,360 @@
+ - Relative path given to require by user: `"./sibling"`
+
+**Approach 1: Translate to the "correct" relative path**
+ - Translated relative path given to `fopen`/`_wfopen`: `"../../sibling"`
+
+**Approach 2: Convert the given relative path into its corresponding absolute path**
+ - Translated relative path given to `fopen`/`_wfopen`: `"/Users/johndoe/project/sibling"`
+
+Although `fopen`/`_wfopen` can handle both relative (to CWD) and absolute paths, the second approach makes more sense for our use case. We already need absolute paths for caching, as explained in the "Caching" section, so we might as well generate these absolute paths during the path resolution stage. With the first approach, we would need to call `realpath` to convert the relative-to-CWD path into an absolute path for caching, which is an unnecessary extra step.
+
+The second approach is also easier to implement. In order to implement the first approach, we would need to constantly keep track of the relationship between the current working directory and the requiring file in order to translate paths correctly. With the second approach, we can safely ignore the location of the current working directory: everything will be relative to the absolute path of the requiring file.
+
+Using the second approach for the example above, we know that the absolute path to the requiring file is `"/Users/johndoe/project/requirer.luau`. To find `"./sibling"`, we can simply traverse the requiring file's absolute path to arrive at `"/Users/johndoe/project/sibling.luau`. This logic is fairly easy to implement (for example, if `".."`, move up one directory).
+
+#### Where to begin traversal
+
+One key assumption of this section is that we will have the absolute path of the requiring file when requiring a module by relative path.
+
+While we could add an explicit reference to this directory to the `lua_State`, we already have an internal mechanism that allows us to get this information. We essentially want to call the `C++` equivalent of `debug.info(1, "s")` when we enter `lua_require`, which would return the name of the file that called `require`, or `stdin` if the module was required directly from the CLI.
+
+As an example, we might do something like this in `lua_require`:
+
+```cpp
+static int lua_require(lua_State* L)
+{
+    lua_Debug ar;
+    lua_getinfo(L, 1, "s", &ar);
+
+    // Path of requiring file
+    const char* basepath = ar.source;
+
+    // ...
+}
+```
+
+#### Impact on `debug.info` output
+
+In order for the above approach (using `debug.info` to obtain the requiring file's path) to work properly, we need to change the `chunkname` that is passed to `luau_load` [here](https://github.com/Roblox/luau/blob/e25de95445f2d635a125ab463426bb7fda017093/CLI/Repl.cpp#L152) to be an absolute path. This slightly changes the behavior of `debug.info`: the source file name returned by `debug.info` will now be an absolute path to the file rather than the string that was used to require that file.
```

Comment:
Hmm. This gets us into somewhat complicated territory. This means that scripts will be aware of absolute path, including eg the user name for cases when script files are part of user home directory. This also means that we get exposed to a bunch of Unicode and path compatibility issues.

Reply:
Does storing cached modules using their absolute paths as keys also have the same issue? Would we need to come up with another system for coming up with unique keys (e.g. pass the absolute path through a hash function)?

Reply:
I think this is only a problem if the absolute path ends up getting exposed to scripts. If our require implementation knows where to get the absolute path from but doesn't expose it to the script then there's no issue in either case (well, Unicode path issues still exist but these are more minor).

Reply:
We could keep the same system for chunknames that we currently have (just prepend the user's given string with '=') and resolve it to an absolute path in `lua_require`. 

This has its own issues though... in the current implementation, everything is relative to the CWD, so when `debug.info` spits out something like `../../../module`, at least it makes sense within that context. However, since we're proposing no longer being relative to the CWD, the "user-given" path might actually be provided by a requiring file that is located somewhere other than the CWD. Using this as-is would likely make the output of `debug.info` pretty confusing.

We had discussed this in the past, but rather than doing everything in terms of absolute paths under the hood, we could keep doing everything in terms of the CWD, making sure to correctly translate a relative path that is relative to the requiring file to be relative to the CWD internally. This would be [Approach 1](https://github.com/VarunSaini02/luau/blob/667a42d2168d70b923b800adc99d37794a5d9cc9/rfcs/require-by-string.md?plain=1#L238) that I had added to the implementation details section.

Reply:
I guess there's a fundamental question, which is that if a module gets referenced by two different aliases from two different libraries, what is the path to that module as reported by `debug.info`. The only stable path is the absolute path, but disclosing that to the script feels bad... The CWD-relative path is a good idea in that normally you have a notion of the "root" of the project, that all paths can be relative to. The one downside here is that on Windows, relative paths can't cross drive letter boundaries....

Reply:
Here's what I'm thinking regarding that. We know that someone who does want to cross the drive letter boundary on Windows has to use an absolute path in some way. Based on our more recent discussion about absolute paths in `require` statements, the only way to do this is through an alias. So, what if we stored script-accessible chunk names like this:

- By default, transform the given path to be relative to the CWD, and use this as the chunk name.
- If the given path contains an alias, keep the alias in the chunk name (i.e. `debug.info` would return something like `@Library/module`)

Reply:
I've described this approach in this commit. [3ef99e2](https://github.com/Roblox/luau/pull/969/commits/3ef99e22cef5c7ce0d9d28fad805985c7a4cbd48)